numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -0,0 +1,245 @@
+    char *tp_name = PyDataMem_NEW(100);
+    snprintf(tp_name, 100, "numpy.dtype[%s]",
+             descr->typeobj->tp_name);
+
+    PyArray_DTypeMeta *dtype_class = PyDataMem_NEW(sizeof(PyArray_DTypeMeta));
+    if (dtype_class == NULL) {
+        return -1;
+    }
+    /*
+     * Initialize the struct fields similar to static code:
+     */
+    /* Copy in np.dtype since it shares most things */
+    memcpy(dtype_class, &PyArrayDescr_Type, sizeof(PyArray_DTypeMeta));
+
+    /* Fix name, base, and __new__*/
+    ((PyTypeObject *)dtype_class)->tp_name = tp_name;
+    ((PyTypeObject *)dtype_class)->tp_base = &PyArrayDescr_Type;
+    ((PyTypeObject *)dtype_class)->tp_new = (newfunc)legacy_dtype_default_new;
+
+    /* Let python finish the initialization (probably unnecessary) */
+    if (PyType_Ready((PyTypeObject *)dtype_class) < 0) {
+        return -1;
+    }
+
+    /* np.dtype is not a concrete DType, this one is */
+    dtype_class->is_abstract = NPY_FALSE;
+
+    Py_INCREF(descr);  /* descr is a singleton that must survive, ensure. */
+    dtype_class->singleton = descr;
+    Py_INCREF(descr->typeobj);
+    dtype_class->scalar_type = descr->typeobj;
+    dtype_class->is_legacy = NPY_TRUE;
+    dtype_class->type_num = descr->type_num;
+    dtype_class->type = descr->type;
+    dtype_class->f = descr->f;
+    dtype_class->kind = descr->kind;
+    dtype_class->itemsize = descr->elsize;
+
+    if (PyTypeNum_ISDATETIME(descr->type_num)) {
+        /* Datetimes are flexible, but were not considered previously */
+        dtype_class->is_flexible = NPY_TRUE;
+    }
+    if (PyTypeNum_ISFLEXIBLE(descr->type_num)) {
+        dtype_class->is_flexible = NPY_TRUE;
+        dtype_class->itemsize = -1;  /* itemsize is not fixed */
+    }
+
+    /* Finally, replace the current class of the descr */
+    Py_TYPE(descr) = (PyTypeObject *)dtype_class;
+
+    return 0;
+}
+
+
+/*
+ * Simple exposed information, defined for each DType (class). This is
+ * preliminary (the flags should also return bools).
+ */
+static PyMemberDef dtypemeta_members[] = {
+        {"_abstract",
+                T_BYTE, offsetof(PyArray_DTypeMeta, is_abstract), READONLY, NULL},
+        {"type",
+                T_OBJECT, offsetof(PyArray_DTypeMeta, scalar_type), READONLY, NULL},
+        {"_flexible",
+                T_BYTE, offsetof(PyArray_DTypeMeta, is_flexible), READONLY, NULL},
+        {NULL, 0, 0, 0, NULL},
+};
+
+
+NPY_NO_EXPORT PyTypeObject PyArrayDTypeMeta_Type = {
+        PyVarObject_HEAD_INIT(NULL, 0)
+        .tp_name = "numpy._DTypeMeta",
+        .tp_basicsize = sizeof(PyArray_DTypeMeta),
+        /* methods */
+        .tp_dealloc = (destructor)dtypemeta_dealloc,
+        /* Types are garbage collected (see dtypemeta_is_gc documentation) */
+        .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+        .tp_doc = "Preliminary NumPy API: The Type of NumPy DTypes (metaclass)",
+        .tp_members = dtypemeta_members,
+        .tp_base = &PyType_Type,
```

Comment:
Ah, this (and similar in `np.dtype`) is not strictly valid, and I have to do that during module loading. Maybe the reason for the above not working...